File: golang/direct_e2ee/v2_wire_vectors_test.go
package directe2ee

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"
	"time"

	anp "github.com/agent-network-protocol/anp/golang"
	"github.com/agent-network-protocol/anp/golang/authentication"
	"github.com/agent-network-protocol/anp/golang/proof"
)

func loadV2Vectors(t *testing.T) map[string]any {
	t.Helper()
	path := filepath.Join("..", "..", "testdata", "direct_e2ee", "p5_v2_wire_vectors.json")
	contents, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	var result map[string]any
	if err := json.Unmarshal(contents, &result); err != nil {
		t.Fatal(err)
	}
	return result
}

func TestV2SharedBundleAndRPCVectors(t *testing.T) {
	vectors := loadV2Vectors(t)
	var bundle V2PrekeyBundle
	if err := decodeV2(vectors["prekey_bundle"], &bundle); err != nil {
		t.Fatal(err)
	}
	canonical, err := SignedBundleObjectJCSV2(bundle)
	if err != nil {
		t.Fatal(err)
	}
	if string(canonical) != vectors["expected_signed_bundle_object_jcs"].(string) {
		t.Fatalf("bundle JCS mismatch: %s", canonical)
	}

	meta, body, err := ParsePublishPrekeyBundleRequestV2(vectors["publish_request"])
	if err != nil {
		t.Fatal(err)
	}
	rebuilt, err := PublishPrekeyBundleRequestV2(meta, body)
	if err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(rebuilt, vectors["publish_request"]) {
		t.Fatalf("publish request mismatch")
	}

	getMeta, getBody, err := ParseGetPrekeyBundleRequestV2(vectors["get_request"])
	if err != nil {
		t.Fatal(err)
	}
	rebuilt, err = GetPrekeyBundleRequestV2(getMeta, getBody)
	if err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(rebuilt, vectors["get_request"]) {
		t.Fatalf("get request mismatch")
	}

	if _, err := ParsePublishPrekeyBundleResultV2(vectors["publish_result"]); err != nil {
		t.Fatal(err)
	}
	if _, err := ParseGetPrekeyBundleResultV2(vectors["get_result"]); err != nil {
		t.Fatal(err)
	}
	if _, err := ParseDirectSendResultV2(vectors["direct_send_result"]); err != nil {
		t.Fatal(err)
	}

	invalidGet := cloneV2Value(t, vectors["get_result"])
	invalidGet["target_device_id"] = "dev-sibling"
	if _, err := ParseGetPrekeyBundleResultV2(invalidGet); err == nil {
		t.Fatal("accepted mismatched get result device")
	}
	invalidDirect := cloneV2Value(t, vectors["direct_send_result"])
	invalidDirect["operation_id"] = "different-operation"
	if _, err := ParseDirectSendResultV2(invalidDirect); err == nil {
		t.Fatal("accepted mismatched direct.send result identifiers")
	}
	invalidPublish := cloneV2Value(t, vectors["publish_result"])
	invalidPublish["unexpected"] = true
	if _, err := ParsePublishPrekeyBundleResultV2(invalidPublish); err == nil {
		t.Fatal("accepted unknown publish result field")
	}
}

func TestV2SharedSignedBundleGoldenVerifies(t *testing.T) {
	vectors := loadV2Vectors(t)
	golden := vectors["signed_bundle_golden"].(map[string]any)
	var bundle V2PrekeyBundle
	if err := decodeV2(golden["prekey_bundle"], &bundle); err != nil {
		t.Fatal(err)
	}
	var didDocument map[string]any
	encoded, _ := json.Marshal(golden["did_document"])
	if err := json.Unmarshal(encoded, &didDocument); err != nil {
		t.Fatal(err)
	}
	now, err := time.Parse(time.RFC3339, golden["now"].(string))
	if err != nil {
		t.Fatal(err)
	}
	if err := VerifyPrekeyBundleV2(bundle, didDocument, now); err != nil {
		t.Fatalf("cross-language signed bundle failed: %v", err)
	}
	bundle.SignedPrekey.KeyID = "spk-tampered"
	if err := VerifyPrekeyBundleV2(bundle, didDocument, now); err == nil {
		t.Fatal("accepted tampered signed bundle")
	}
}

func cloneV2Value(t *testing.T, value any) map[string]any {
	t.Helper()
	encoded, err := json.Marshal(value)
	if err != nil {
		t.Fatal(err)
	}
	var result map[string]any
	if err := json.Unmarshal(encoded, &result); err != nil {
		t.Fatal(err)
	}
	return result
}

func TestV2SharedAADAndPlaintextVectors(t *testing.T) {
	vectors := loadV2Vectors(t)
	meta, rawBody, err := ParseDirectSendRequestV2(vectors["direct_init_request"])
	if err != nil {
		t.Fatal(err)
	}
	initBody := rawBody.(V2DirectInitBody)
	aad, err := BuildInitAADV2(meta, initBody)
	if err != nil {
		t.Fatal(err)
	}
	if string(aad) != vectors["expected_ad_init"].(string) {
		t.Fatalf("AD_init mismatch: %s", aad)
	}
	rebuilt, err := DirectSendRequestV2(meta, initBody)
	if err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(rebuilt, vectors["direct_init_request"]) {
		t.Fatal("init request mismatch")
	}

	meta, rawBody, err = ParseDirectSendRequestV2(vectors["direct_cipher_request"])
	if err != nil {
		t.Fatal(err)
	}
	cipherBody := rawBody.(V2DirectCipherBody)
	aad, err = BuildMessageAADV2(meta, cipherBody)
	if err != nil {
		t.Fatal(err)
	}
	if string(aad) != vectors["expected_ad_msg"].(string) {
		t.Fatalf("AD_msg mismatch: %s", aad)
	}

	var plaintext V2ApplicationPlaintext
	if err := decodeV2(vectors["application_plaintext"], &plaintext); err != nil {
		t.Fatal(err)
	}
	plain, err := CanonicalApplicationPlaintextV2(plaintext)
	if err != nil {
		t.Fatal(err)
	}
	if string(plain) != vectors["expected_application_plaintext_jcs"].(string) {
		t.Fatalf("plaintext JCS mismatch: %s", plain)
	}
	if err := decodeV2(vectors["application_plaintext_numeric"], &plaintext); err != nil {
		t.Fatal(err)
	}
	plain, err = CanonicalApplicationPlaintextV2(plaintext)
	if err != nil {
		t.Fatal(err)
	}
	if string(plain) != vectors["expected_application_plaintext_numeric_jcs"].(string) {
		t.Fatalf("numeric plaintext JCS mismatch: %s", plain)
	}
}

func TestV2DeviceTamperAndForbiddenFields(t *testing.T) {
	vectors := loadV2Vectors(t)
	clone := func() map[string]any {
		encoded, _ := json.Marshal(vectors["direct_init_request"])
		var result map[string]any
		_ = json.Unmarshal(encoded, &result)
		return result
	}
	tampered := clone()
	tampered["params"].(map[string]any)["meta"].(map[string]any)["recipient_device_id"] = "dev-sibling"
	meta, rawBody, err := ParseDirectSendRequestV2(tampered)
	if err != nil {
		t.Fatal(err)
	}
	aad, err := BuildInitAADV2(meta, rawBody.(V2DirectInitBody))
	if err != nil {
		t.Fatal(err)
	}
	if string(aad) == vectors["expected_ad_init"].(string) {
		t.Fatal("device tamper did not change AAD")
	}
	tampered = clone()
	tampered["params"].(map[string]any)["meta"].(map[string]any)["sender_device_id"] = "dev-sender-sibling"
	meta, rawBody, err = ParseDirectSendRequestV2(tampered)
	if err != nil {
		t.Fatal(err)
	}
	aad, err = BuildInitAADV2(meta, rawBody.(V2DirectInitBody))
	if err != nil {
		t.Fatal(err)
	}
	if string(aad) == vectors["expected_ad_init"].(string) {
		t.Fatal("sender device tamper did not change AAD")
	}

	for _, field := range []string{"auth", "deliveries", "root_private_key", "document_version"} {
		invalid := clone()
		invalid["params"].(map[string]any)[field] = map[string]any{}
		if _, _, err := ParseDirectSendRequestV2(invalid); err == nil {
			t.Fatalf("accepted forbidden field %s", field)
		}
	}
	invalid := clone()
	invalid["params"].(map[string]any)["meta"].(map[string]any)["logical_message_id"] = "outer"
	if _, _, err := ParseDirectSendRequestV2(invalid); err == nil {
		t.Fatal("accepted outer logical_message_id")
	}
	invalid = clone()
	invalid["params"].(map[string]any)["meta"].(map[string]any)["operation_id"] = "different-id"
	if _, _, err := ParseDirectSendRequestV2(invalid); err == nil {
		t.Fatal("accepted operation_id != message_id")
	}
}

func TestV2PublishRejectsEmptyOPKsAndBadPublicKey(t *testing.T) {
	vectors := loadV2Vectors(t)
	clone := func() map[string]any {
		encoded, _ := json.Marshal(vectors["publish_request"])
		var result map[string]any
		_ = json.Unmarshal(encoded, &result)
		return result
	}
	invalid := clone()
	invalid["params"].(map[string]any)["body"].(map[string]any)["one_time_prekeys"] = []any{}
	if _, _, err := ParsePublishPrekeyBundleRequestV2(invalid); err == nil {
		t.Fatal("accepted explicit empty one_time_prekeys")
	}
	invalid = clone()
	invalid["params"].(map[string]any)["body"].(map[string]any)["prekey_bundle"].(map[string]any)["signed_prekey"].(map[string]any)["public_key_b64u"] = "AA=="
	if _, _, err := ParsePublishPrekeyBundleRequestV2(invalid); err == nil {
		t.Fatal("accepted padded/non-X25519 public key")
	}
}

func TestV2SharedInvalidWireEncodings(t *testing.T) {
	vectors := loadV2Vectors(t)
	invalid := vectors["encoding_negative_values"].(map[string]any)

	request := cloneV2Value(t, vectors["direct_init_request"])
	request["params"].(map[string]any)["body"].(map[string]any)["session_id"] = invalid["session_id"]
	if _, _, err := ParseDirectSendRequestV2(request); err == nil {
		t.Fatal("accepted invalid session_id")
	}
	request = cloneV2Value(t, vectors["direct_init_request"])
	request["params"].(map[string]any)["body"].(map[string]any)["sender_ephemeral_pub_b64u"] = invalid["x25519_public_key"]
	if _, _, err := ParseDirectSendRequestV2(request); err == nil {
		t.Fatal("accepted invalid sender ephemeral key")
	}
	request = cloneV2Value(t, vectors["direct_init_request"])
	request["params"].(map[string]any)["body"].(map[string]any)["ciphertext_b64u"] = invalid["ciphertext_b64u"]
	if _, _, err := ParseDirectSendRequestV2(request); err == nil {
		t.Fatal("accepted invalid ciphertext")
	}
	request = cloneV2Value(t, vectors["direct_cipher_request"])
	request["params"].(map[string]any)["body"].(map[string]any)["ratchet_header"].(map[string]any)["dh_pub_b64u"] = invalid["x25519_public_key"]
	if _, _, err := ParseDirectSendRequestV2(request); err == nil {
		t.Fatal("accepted invalid ratchet public key")
	}
	request = cloneV2Value(t, vectors["direct_cipher_request"])
	request["params"].(map[string]any)["meta"].(map[string]any)["created_at"] = invalid["created_at"]
	if _, _, err := ParseDirectSendRequestV2(request); err == nil {
		t.Fatal("accepted invalid created_at")
	}
	payload := invalid["payload_b64u"].(string)
	plaintext := V2ApplicationPlaintext{
		ApplicationContentType: "application/octet-stream",
		PayloadB64U:            &payload,
	}
	if err := plaintext.Validate(); err == nil {
		t.Fatal("accepted invalid payload_b64u")
	}
}

func TestV2OptionalMetaDoesNotEnterAADAndErrorsAreExact(t *testing.T) {
	vectors := loadV2Vectors(t)
	meta, rawBody, err := ParseDirectSendRequestV2(vectors["direct_init_request"])
	if err != nil {
		t.Fatal(err)
	}
	original, _ := BuildInitAADV2(meta, rawBody.(V2DirectInitBody))
	meta.ANPVersion = "9.9"
	meta.CreatedAt = "2030-01-01T00:00:00Z"
	changed, _ := BuildInitAADV2(meta, rawBody.(V2DirectInitBody))
	if string(original) != string(changed) {
		t.Fatal("optional outer meta entered AAD")
	}

	errors := vectors["errors"].([]any)
	if len(errors) != len(V2ProtocolErrors) {
		t.Fatal("error table length mismatch")
	}
	for _, raw := range errors {
		expected := raw.(map[string]any)
		entry, ok := DirectE2EEV2ErrorByCode(int(expected["code"].(float64)))
		if !ok || entry.ANPCode != expected["anp_code"].(string) {
			t.Fatalf("error table mismatch: %#v", expected)
		}
	}
	if _, ok := DirectE2EEV2ErrorByCode(5000); ok {
		t.Fatal("accepted non-P5 code")
	}
}

func TestV2RejectsExplicitNullsAndContentBearerMismatches(t *testing.T) {
	vectors := loadV2Vectors(t)
	request := cloneV2Value(t, vectors["direct_init_request"])
	request["params"].(map[string]any)["body"].(map[string]any)["recipient_one_time_prekey_id"] = nil
	if _, _, err := ParseDirectSendRequestV2(request); err == nil {
		t.Fatal("accepted null recipient_one_time_prekey_id")
	}
	for _, field := range []string{"preferred_suite", "require_opk"} {
		request = cloneV2Value(t, vectors["get_request"])
		request["params"].(map[string]any)["body"].(map[string]any)[field] = nil
		if _, _, err := ParseGetPrekeyBundleRequestV2(request); err == nil {
			t.Fatalf("accepted null %s", field)
		}
	}

	invalidPlaintexts := []map[string]any{
		{"application_content_type": "text/plain", "payload": map[string]any{}},
		{"application_content_type": "application/json", "text": "wrong"},
		{"application_content_type": "application/json", "annotations": []any{}, "payload": map[string]any{}},
		{"application_content_type": "application/json", "annotations": nil, "payload": map[string]any{}},
	}
	for _, raw := range invalidPlaintexts {
		var plaintext V2ApplicationPlaintext
		if err := decodeV2(raw, &plaintext); err == nil {
			if err := plaintext.Validate(); err == nil {
				t.Fatalf("accepted invalid plaintext: %#v", raw)
			}
		}
	}
}

func TestV2BundleObjectProofCoversDeviceAndStaticFields(t *testing.T) {
	generated, err := authentication.CreateDidWBADocument("bundle-v2.example", authentication.DidDocumentOptions{
		PathSegments: []string{"agents", "alice"}, DidProfile: authentication.DidProfileE1,
	})
	if err != nil {
		t.Fatal(err)
	}
	document := generated.DidDocument
	did := document["id"].(string)
	delete(document, "proof")
	document["deviceManifest"] = map[string]any{
		"type": "ANPDeviceManifest",
		"devices": []any{map[string]any{
			"device_id": "dev-a", "signing_key_id": did + "#key-1",
			"e2ee_key_id": did + "#key-3",
			"profiles":    []any{"anp.core.binding.v2", "anp.identity.discovery.v2", "anp.direct.base.v2", "anp.direct.e2ee.v2"},
		}},
	}
	signingKey, err := anp.PrivateKeyFromPEM(generated.Keys[authentication.VMKeyAuth].PrivateKeyPEM)
	if err != nil {
		t.Fatal(err)
	}
	document, err = proof.GenerateW3CProof(document, signingKey, did+"#key-1", proof.GenerationOptions{
		ProofPurpose: "assertionMethod", ProofType: proof.ProofTypeDataIntegrity,
		Cryptosuite: proof.CryptosuiteEddsaJCS2022, Created: "2026-07-19T00:00:00Z",
	})
	if err != nil {
		t.Fatal(err)
	}
	bundle, err := BuildPrekeyBundleV2(
		"bundle-v2", did, "dev-a", did+"#key-3",
		V2SignedPrekey{KeyID: "spk-v2", PublicKeyB64U: "UKYUCbHd0DJemxa3AOcZ6XcsBwALG9d4bpB8ZT0gSV0", ExpiresAt: "2035-01-01T00:00:00Z"},
		signingKey, did+"#key-1", "2026-07-19T00:00:00Z",
	)
	if err != nil {
		t.Fatal(err)
	}
	if _, err := BuildPrekeyBundleV2(
		"bundle-invalid", did, "dev-a", did+"#key-3",
		V2SignedPrekey{KeyID: "spk-invalid", PublicKeyB64U: "AA==", ExpiresAt: "not-rfc3339"},
		signingKey, did+"#key-1", "2026-07-19T00:00:00Z",
	); err == nil {
		t.Fatal("builder accepted invalid signed prekey")
	}
	if err := VerifyPrekeyBundleV2(bundle, document, time.Date(2026, 7, 19, 0, 0, 1, 0, time.UTC)); err != nil {
		t.Fatal(err)
	}
	tampered := bundle
	tampered.OwnerDeviceID = "dev-sibling"
	if err := VerifyPrekeyBundleV2(tampered, document, time.Date(2026, 7, 19, 0, 0, 1, 0, time.UTC)); err == nil {
		t.Fatal("owner_device_id tamper passed")
	}
	tampered = bundle
	tampered.StaticKeyAgreementID = did + "#other"
	if err := VerifyPrekeyBundleV2(tampered, document, time.Date(2026, 7, 19, 0, 0, 1, 0, time.UTC)); err == nil {
		t.Fatal("static key tamper passed")
	}
}

File: dart/test/direct_e2ee/v2_wire_vectors_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:anp/anp.dart';
import 'package:test/test.dart';

void main() {
  final vectors = _loadVectors();

  test('shared P5 v2 bundle and key-service requests round trip', () {
    final bundle = V2PrekeyBundle.fromJson(_map(vectors['prekey_bundle']));
    expect(
      signedBundleObjectJcsV2(bundle),
      vectors['expected_signed_bundle_object_jcs'],
    );

    final (publishMeta, publishBundle, opks) =
        parsePublishPrekeyBundleRequestV2(_map(vectors['publish_request']));
    expect(
      publishPrekeyBundleRequestV2(
        publishMeta,
        publishBundle,
        oneTimePrekeys: opks,
      ),
      equals(vectors['publish_request']),
    );

    final (getMeta, getBody) = parseGetPrekeyBundleRequestV2(
      _map(vectors['get_request']),
    );
    expect(
      getPrekeyBundleRequestV2(
        getMeta,
        targetDid: getBody['target_did']! as String,
        targetDeviceId: getBody['target_device_id']! as String,
        preferredSuite: getBody['preferred_suite'] as String?,
        requireOpk: getBody['require_opk'] as bool?,
      ),
      equals(vectors['get_request']),
    );

    expect(
      parsePublishPrekeyBundleResultV2(_map(vectors['publish_result'])),
      isA<V2PublishPrekeyBundleResult>(),
    );
    expect(
      parseGetPrekeyBundleResultV2(_map(vectors['get_result'])),
      isA<V2GetPrekeyBundleResult>(),
    );
    expect(
      parseDirectSendResultV2(_map(vectors['direct_send_result'])),
      isA<V2DirectSendResult>(),
    );

    final invalidGet = _clone(vectors['get_result'])
      ..['target_device_id'] = 'dev-sibling';
    expect(
      () => parseGetPrekeyBundleResultV2(invalidGet),
      throwsA(isA<AnpDirectE2eeV2Exception>()),
    );
    final invalidDirect = _clone(vectors['direct_send_result'])
      ..['operation_id'] = 'different-operation';
    expect(
      () => parseDirectSendResultV2(invalidDirect),
      throwsA(isA<AnpDirectE2eeV2Exception>()),
    );
    final invalidPublish = _clone(vectors['publish_result'])
      ..['unexpected'] = true;
    expect(
      () => parsePublishPrekeyBundleResultV2(invalidPublish),
      throwsA(isA<AnpDirectE2eeV2Exception>()),
    );
  });

  test('shared P5 v2 AAD and ApplicationPlaintext vectors match', () {
    final (initMeta, rawInit) = parseDirectSendRequestV2(
      _map(vectors['direct_init_request']),
    );
    final init = rawInit as V2DirectInitBody;
    expect(
      utf8.decode(buildInitAadV2(initMeta, init)),
      vectors['expected_ad_init'],
    );
    expect(
      directSendRequestV2(initMeta, init),
      equals(vectors['direct_init_request']),
    );

    final (cipherMeta, rawCipher) = parseDirectSendRequestV2(
      _map(vectors['direct_cipher_request']),
    );
    final cipher = rawCipher as V2DirectCipherBody;
    expect(
      utf8.decode(buildMessageAadV2(cipherMeta, cipher)),
      vectors['expected_ad_msg'],
    );
    final plaintext = V2ApplicationPlaintext.fromJson(
      _map(vectors['application_plaintext']),
    );
    expect(
      utf8.decode(canonicalApplicationPlaintextV2(plaintext)),
      vectors['expected_application_plaintext_jcs'],
    );
    final numeric = V2ApplicationPlaintext.fromJson(
      _map(vectors['application_plaintext_numeric']),
    );
    expect(
      utf8.decode(canonicalApplicationPlaintextV2(numeric)),
      vectors['expected_application_plaintext_numeric_jcs'],
    );
  });

  test('device tamper changes AAD and private/aggregate fields fail', () {
    final request = _clone(vectors['direct_init_request']);
    final params = _map(request['params']);
    final meta = _map(params['meta']);
    meta['recipient_device_id'] = 'dev-sibling';
    params['meta'] = meta;
    request['params'] = params;
    final (parsedMeta, body) = parseDirectSendRequestV2(request);
    expect(
      utf8.decode(buildInitAadV2(parsedMeta, body as V2DirectInitBody)),
      isNot(vectors['expected_ad_init']),
    );

    final senderTampered = _clone(vectors['direct_init_request']);
    final senderParams = _map(senderTampered['params']);
    final senderMeta = _map(senderParams['meta']);
    senderMeta['sender_device_id'] = 'dev-sender-sibling';
    senderParams['meta'] = senderMeta;
    senderTampered['params'] = senderParams;
    final (parsedSenderMeta, senderBody) = parseDirectSendRequestV2(
      senderTampered,
    );
    expect(
      utf8.decode(
        buildInitAadV2(parsedSenderMeta, senderBody as V2DirectInitBody),
      ),
      isNot(vectors['expected_ad_init']),
    );

    for (final field in [
      'auth',
      'deliveries',
      'root_private_key',
      'document_version',
    ]) {
      final invalid = _clone(vectors['direct_init_request']);
      final invalidParams = _map(invalid['params'])
        ..[field] = <String, Object?>{};
      invalid['params'] = invalidParams;
      expect(
        () => parseDirectSendRequestV2(invalid),
        throwsA(isA<AnpDirectE2eeV2Exception>()),
      );
    }
    final invalid = _clone(vectors['direct_init_request']);
    final invalidParams = _map(invalid['params']);
    final invalidMeta = _map(invalidParams['meta'])
      ..['logical_message_id'] = 'outer-logical';
    invalidParams['meta'] = invalidMeta;
    invalid['params'] = invalidParams;
    expect(
      () => parseDirectSendRequestV2(invalid),
      throwsA(isA<AnpDirectE2eeV2Exception>()),
    );

    final operationMismatch = _clone(vectors['direct_init_request']);
    final mismatchParams = _map(operationMismatch['params']);
    final mismatchMeta = _map(mismatchParams['meta'])
      ..['operation_id'] = 'different-operation';
    mismatchParams['meta'] = mismatchMeta;
    operationMismatch['params'] = mismatchParams;
    expect(
      () => parseDirectSendRequestV2(operationMismatch),
      throwsA(isA<AnpDirectE2eeV2Exception>()),
    );
  });

  test('publish rejects empty OPK arrays and invalid X25519 key encodings', () {
    final emptyOpks = _clone(vectors['publish_request']);
    final emptyParams = _map(emptyOpks['params']);
    final emptyBody = _map(emptyParams['body'])
      ..['one_time_prekeys'] = <Object?>[];
    emptyParams['body'] = emptyBody;
    emptyOpks['params'] = emptyParams;
    expect(
      () => parsePublishPrekeyBundleRequestV2(emptyOpks),
      throwsA(isA<AnpDirectE2eeV2Exception>()),
    );

    final badKey = _clone(vectors['publish_request']);
    final badParams = _map(badKey['params']);
    final badBody = _map(badParams['body']);
    final badBundle = _map(badBody['prekey_bundle']);
    final badSignedPrekey = _map(badBundle['signed_prekey'])
      ..['public_key_b64u'] = 'AA==';
    badBundle['signed_prekey'] = badSignedPrekey;
    badBody['prekey_bundle'] = badBundle;
    badParams['body'] = badBody;
    badKey['params'] = badParams;
    expect(
      () => parsePublishPrekeyBundleRequestV2(badKey),
      throwsA(isA<AnpDirectE2eeV2Exception>()),
    );

    final (meta, bundle, _) = parsePublishPrekeyBundleRequestV2(
      _map(vectors['publish_request']),
    );
    final invalidOpk = V2OneTimePrekey(
      keyId: 'opk-invalid',
      publicKeyB64u: 'AA==',
    );
    expect(
      () => publishPrekeyBundleRequestV2(
        meta,
        bundle,
        oneTimePrekeys: [invalidOpk],
      ),
      throwsA(isA<AnpDirectE2eeV2Exception>()),
    );
  });

  test('shared invalid P5 wire encodings are rejected', () {
    final invalidValues = _map(vectors['encoding_negative_values']);

    JsonMap invalidRequest(
      String requestName,
      void Function(JsonMap params) change,
    ) {
      final request = _clone(vectors[requestName]);
      final params = _map(request['params']);
      change(params);
      request['params'] = params;
      return request;
    }

    final requests = <JsonMap>[
      invalidRequest('direct_init_request', (params) {
        final body = _map(params['body'])
          ..['session_id'] = invalidValues['session_id'];
        params['body'] = body;
      }),
      invalidRequest('direct_init_request', (params) {
        final body = _map(params['body'])
          ..['sender_ephemeral_pub_b64u'] = invalidValues['x25519_public_key'];
        params['body'] = body;
      }),
      invalidRequest('direct_init_request', (params) {
        final body = _map(params['body'])
          ..['ciphertext_b64u'] = invalidValues['ciphertext_b64u'];
        params['body'] = body;
      }),
      invalidRequest('direct_cipher_request', (params) {
        final body = _map(params['body']);
        final header = _map(body['ratchet_header'])
          ..['dh_pub_b64u'] = invalidValues['x25519_public_key'];
        body['ratchet_header'] = header;
        params['body'] = body;
      }),
      invalidRequest('direct_cipher_request', (params) {
        final meta = _map(params['meta'])
          ..['created_at'] = invalidValues['created_at'];
        params['meta'] = meta;
      }),
    ];
    for (final request in requests) {
      expect(
        () => parseDirectSendRequestV2(request),
        throwsA(isA<AnpDirectE2eeV2Exception>()),
      );
    }
    expect(
      () => V2ApplicationPlaintext.fromJson({
        'application_content_type': 'application/octet-stream',
        'payload_b64u': invalidValues['payload_b64u'],
      }),
      throwsA(isA<AnpDirectE2eeV2Exception>()),
    );
  });

  test('optional outer fields do not enter AAD and errors are exact', () {
    final (meta, body) = parseDirectSendRequestV2(
      _map(vectors['direct_init_request']),
    );
    final original = buildInitAadV2(meta, body as V2DirectInitBody);
    final changed = buildInitAadV2(
      meta.copyWith(anpVersion: '9.9', createdAt: '2030-01-01T00:00:00Z'),
      body,
    );
    expect(changed, original);

    final errors = vectors['errors']! as List<Object?>;
    expect(directE2eeV2Errors, hasLength(13));
    for (final raw in errors) {
      final expected = _map(raw);
      expect(
        directE2eeV2Error(expected['code']! as int)?.anpCode,
        expected['anp_code'],
      );
    }
    expect(directE2eeV2Error(5000), isNull);
  });

  test('bundle binding selects the exact Manifest device and proof key', () {
    final bundle = V2PrekeyBundle.fromJson(_map(vectors['prekey_bundle']));
    final did = bundle.ownerDid;
    final document = <String, Object?>{
      'id': did,
      'verificationMethod': [
        {
          'id': '$did#dev-b-sign',
          'type': 'Multikey',
          'controller': did,
          'publicKeyMultibase': 'zSign',
        },
        {
          'id': '$did#dev-b-e2ee',
          'type': 'Multikey',
          'controller': did,
          'publicKeyMultibase': 'zE2ee',
        },
      ],
      'assertionMethod': ['$did#dev-b-sign'],
      'keyAgreement': ['$did#dev-b-e2ee'],
      'deviceManifest': {
        'type': 'ANPDeviceManifest',
        'devices': [
          {
            'device_id': bundle.ownerDeviceId,
            'signing_key_id': '$did#dev-b-sign',
            'e2ee_key_id': '$did#dev-b-e2ee',
            'profiles': [
              'anp.core.binding.v2',
              'anp.identity.discovery.v2',
              'anp.direct.base.v2',
              'anp.direct.e2ee.v2',
            ],
          },
        ],
      },
    };
    validatePrekeyBundleV2Binding(
      bundle,
      document,
      now: DateTime.utc(2026, 7, 19),
    );
    final tampered = bundle.toJson()..['owner_device_id'] = 'dev-sibling';
    expect(
      () => validatePrekeyBundleV2Binding(
        V2PrekeyBundle.fromJson(tampered),
        document,
        now: DateTime.utc(2026, 7, 19),
      ),
      throwsA(isA<AnpDirectE2eeV2Exception>()),
    );
  });

  test('shared cross-language signed Bundle proof verifies', () async {
    final golden = _map(vectors['signed_bundle_golden']);
    final bundle = V2PrekeyBundle.fromJson(_map(golden['prekey_bundle']));
    final didDocument = _map(golden['did_document']);
    final keyId = bundle.proof['verificationMethod']! as String;
    final method = findVerificationMethod(didDocument, keyId)!;
    final verifier = PublicKeyMessageVerifier({
      keyId: extractPublicKey(method),
    });
    await verifyPrekeyBundleV2(
      bundle,
      didDocument,
      verifier,
      now: DateTime.parse(golden['now']! as String),
    );

    final tampered = bundle.toJson();
    final signedPrekey = _map(tampered['signed_prekey'])
      ..['key_id'] = 'spk-tampered';
    tampered['signed_prekey'] = signedPrekey;
    await expectLater(
      () => verifyPrekeyBundleV2(
        V2PrekeyBundle.fromJson(tampered),
        didDocument,
        verifier,
        now: DateTime.parse(golden['now']! as String),
      ),
      throwsA(isA<AnpDirectE2eeV2Exception>()),
    );
  });

  test('rejects explicit nulls and content bearer mismatches', () {
    final nullOpk = _clone(vectors['direct_init_request']);
    final nullOpkParams = _map(nullOpk['params']);
    final nullOpkBody = _map(nullOpkParams['body'])
      ..['recipient_one_time_prekey_id'] = null;
    nullOpk['params'] = {...nullOpkParams, 'body': nullOpkBody};
    expect(
      () => parseDirectSendRequestV2(nullOpk),
      throwsA(isA<AnpDirectE2eeV2Exception>()),
    );

    for (final field in ['preferred_suite', 'require_opk']) {
      final request = _clone(vectors['get_request']);
      final params = _map(request['params']);
      final body = _map(params['body'])..[field] = null;
      request['params'] = {...params, 'body': body};
      expect(
        () => parseGetPrekeyBundleRequestV2(request),
        throwsA(isA<AnpDirectE2eeV2Exception>()),
      );
    }

    for (final plaintext in <JsonMap>[
      {
        'application_content_type': 'text/plain',
        'payload': <String, Object?>{},
      },
      {'application_content_type': 'application/json', 'text': 'wrong'},
      {
        'application_content_type': 'application/json',
        'annotations': <Object?>[],
        'payload': <String, Object?>{},
      },
      {
        'application_content_type': 'application/json',
        'annotations': null,
        'payload': <String, Object?>{},
      },
    ]) {
      expect(
        () => V2ApplicationPlaintext.fromJson(plaintext),
        throwsA(isA<AnpDirectE2eeV2Exception>()),
      );
    }
  });
}

JsonMap _loadVectors() {
  final path = File('../testdata/direct_e2ee/p5_v2_wire_vectors.json').absolute;
  return _map(jsonDecode(path.readAsStringSync()));
}

JsonMap _clone(Object? value) => _map(jsonDecode(jsonEncode(value)));

JsonMap _map(Object? value) => Map<String, Object?>.from(
  (value! as Map<Object?, Object?>).cast<String, Object?>(),
);
